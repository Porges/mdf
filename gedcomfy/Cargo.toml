[package]
name = "gedcomfy"
version = "0.1.0"
edition.workspace = true
license.workspace = true
authors.workspace = true

[features]
default = ["miette-highlighting", "kdl"]
miette-highlighting = ["miette/fancy-no-backtrace"]
legacy-encodings = ["dep:oem_cp"]
kdl = ["dep:kdl"]

[dependencies]
ascii = "1.1.0"
encoding_rs = "0.8.34"
# TODO: update once this is released https://github.com/zkat/miette/pull/393
miette = { git = "https://github.com/porges/miette", features = ["fancy-no-backtrace"], branch = "inherit-source-to-causes" }
oem_cp = { version = "2.0.0", optional = true }
owo-colors = { version = "4.0.0", features = ["supports-colors"] }
paste = "1.0.15"
thiserror = "1.0.63"
tracing = { version = "0.1", features = ["attributes"] }
vec1 = "1.12.1"
kdl = { version = "4.6.0", optional = true }
yoke = { version = "0.7.4", features = ["derive"] }

[dev-dependencies]
insta = { version = "1.39.0", features = ["glob"] }
