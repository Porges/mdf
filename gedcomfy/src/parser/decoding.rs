use super::{
    encodings::{EncodingError, InvalidDataForEncodingError},
    lines::LineSyntaxError,
    records::RecordStructureError,
    versions::VersionError,
};
use crate::FileStructureError;

#[derive(
    derive_more::Error, derive_more::From, derive_more::Display, Debug, miette::Diagnostic,
)]
pub enum DecodingError {
    #[display("Unable to determine version of GEDCOM file")]
    #[diagnostic(transparent)]
    VersionError(#[from] VersionError),

    #[display("Unable to determine encoding of GEDCOM file")]
    #[diagnostic(transparent)]
    EncodingError(#[from] EncodingError),

    #[display("GEDCOM file contained data which was invalid in the detected encoding")]
    #[diagnostic(transparent)]
    InvalidDataForEncoding(#[from] InvalidDataForEncodingError),

    #[display("GEDCOM file structure is invalid")]
    #[diagnostic(transparent)]
    FileStructureError(#[from] FileStructureError),

    #[display("GEDCOM file contains a record-hierarchy error")]
    #[diagnostic(transparent)]
    RecordStructureError(#[from] RecordStructureError),

    #[display("GEDCOM file contains a syntax error")]
    #[diagnostic(transparent)]
    SyntaxError(#[from] LineSyntaxError),
}
