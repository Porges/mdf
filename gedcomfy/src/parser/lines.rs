use ascii::{AsciiChar, AsciiStr};
use miette::SourceSpan;

use super::{GEDCOMSource, Sourced};

/// Represents a single line in a GEDCOM file.
///
/// Note that we take advantage of the encoding
/// requirements here to make tags less generic,
/// since they must only be part of the ASCII subset.
/// This makes them easier to deal with in code.
pub struct RawLine<'a, S: GEDCOMSource + ?Sized> {
    pub tag: Sourced<&'a AsciiStr>,
    pub xref: Option<Sourced<&'a S>>,
    pub data: Option<Sourced<&'a S>>,
}

/// The types of errors that can occur when parsing lines
/// from a GEDCOM file.
#[derive(thiserror::Error, Debug, miette::Diagnostic)]
pub enum LineSyntaxError {
    #[error("Invalid non-numeric level '{value}'")]
    #[diagnostic(code(gedcom::parse_error::invalid_level))]
    InvalidLevel {
        value: String,
        #[source]
        source: Box<dyn std::error::Error + Send + Sync>,
        #[label("this is not a (positive) number")]
        span: SourceSpan,
    },

    #[error("Reserved value '{reserved_value}' cannot be used as an XRef")]
    #[diagnostic(code(gedcom::parse_error::reserved_xref))]
    ReservedXRef {
        reserved_value: String,
        #[label("{reserved_value} is a reserved value")]
        span: SourceSpan,
    },

    #[error("No tag found")]
    #[diagnostic(code(gedcom::parse_error::no_tag))]
    NoTag {
        #[label("no tag in this line")]
        span: SourceSpan,
    },

    #[error("Invalid character in tag")]
    #[diagnostic(
        code(gedcom::parse_error::invalid_tag),
        help(
            "tag names may only contain the characters a-z, A-Z, and 0-9, or a leading underscore"
        )
    )]
    InvalidTagCharacter {
        #[label("this character is not permitted in a tag")]
        span: SourceSpan,
    },
}

/// This is a straightforward parser for GEDCOM lines. It performs
/// minimal validation, and can be used to parse lines from a
/// ‘decoded’ (`&str`) or ‘raw’ (`&[u8]`) source.
///
/// This is not intended to be used directly by other code, but it
/// may be useful as a basis for other tooling. The `raw` version of
/// this function exists so that records can be parsed in order to determine
/// the encoding of the file before decoding the rest of the file.
///
/// ## Syntax
pub fn iterate_lines<'a, S: GEDCOMSource + ?Sized>(
    source_code: &'a S,
) -> impl Iterator<Item = Result<(Sourced<usize>, Sourced<RawLine<'a, S>>), LineSyntaxError>> {
    // Line syntax is as follows:
    /*
    Line    = Level D [Xref D] Tag [D LineVal] EOL

    Level   = "0" / nonzero *digit
    D       = %x20                            ; space
    Xref    = atsign 1*tagchar atsign         ; but not "@VOID@"
    Tag     = stdTag / extTag
    LineVal = pointer / lineStr
    EOL     = %x0D [%x0A] / %x0A              ; CR-LF, CR, or LF

    stdTag  = ucletter *tagchar
    extTag  = underscore 1*tagchar
    tagchar = ucletter / digit / underscore

    pointer = voidPtr / Xref
    voidPtr = %s"@VOID@"

    nonAt   = %x09 / %x20-3F / %x41-10FFFF    ; non-EOL, non-@
    nonEOL  = %x09 / %x20-10FFFF              ; non-EOL
    lineStr = (nonAt / atsign atsign) *nonEOL ; leading @ doubled
    */

    // TODO: line data should be parsed as pointer|data

    let to_sourced = |s: &'a S| Sourced {
        value: s,
        span: source_code.span_of(s),
    };

    source_code.lines().filter_map(move |line| {
        debug_assert!(!line.ends_with(AsciiChar::LineFeed));
        debug_assert!(!line.ends_with(AsciiChar::CarriageReturn));
        debug_assert!(!line.starts_with(AsciiChar::LineFeed));
        debug_assert!(!line.starts_with(AsciiChar::CarriageReturn));

        let mut parts = line.splitn(4, AsciiChar::Space).peekable();
        let Some(level_part) = parts.next() else {
            unreachable!("even an empty line produces one part")
        };

        if level_part.is_empty() {
            return None; // skipping empty line
        }

        let result = || -> Result<_, _> {
            let level_str = level_part
                .as_ascii_str()
                .map_err(|source| LineSyntaxError::InvalidLevel {
                    source: Box::new(source),
                    value: "<not ascii>".to_string(),
                    span: source_code.span_of(level_part),
                })?
                .as_str();

            let level =
                level_str
                    .parse::<usize>()
                    .map_err(|source| LineSyntaxError::InvalidLevel {
                        source: Box::new(source),
                        value: level_str.to_string(),
                        span: source_code.span_of(level_part),
                    })?;

            let level = Sourced {
                value: level,
                span: source_code.span_of(level_part),
            };

            // XRef starts and ends with '@' but interior does not _have_ to be ASCII
            let xref =
                parts.next_if(|s| s.starts_with(AsciiChar::At) && s.ends_with(AsciiChar::At));

            if let Some(xref) = xref {
                let void = unsafe { AsciiStr::from_ascii_unchecked(b"@VOID@") };
                // tag may not be the reserved 'null' value
                if xref.eq(void) {
                    return Err(LineSyntaxError::ReservedXRef {
                        reserved_value: void.to_string(),
                        span: source_code.span_of(xref),
                    });
                }
            }

            let xref = xref.map(to_sourced);

            let source_tag = parts.next().ok_or_else(|| LineSyntaxError::NoTag {
                span: source_code.span_of(line),
            })?;

            // ensure tag is valid (only ASCII alphanumeric, may have underscore at start)
            let tag = source_tag.as_ascii_str().map_err(|source| {
                // produce error pointing to the first non-valid char
                let full_span = source_code.span_of(source_tag);
                let span = SourceSpan::from((full_span.offset() + source.valid_up_to(), 1));
                LineSyntaxError::InvalidTagCharacter { span }
            })?;

            if let Some((ix, _)) = tag.chars().enumerate().find(|&(ix, char)| {
                if char == AsciiChar::UnderScore {
                    ix > 0
                } else {
                    !char.is_ascii_alphanumeric()
                }
            }) {
                let full_span = source_code.span_of(source_tag);
                let span = SourceSpan::from((full_span.offset() + ix, 1));
                return Err(LineSyntaxError::InvalidTagCharacter { span });
            }

            let tag = Sourced {
                value: tag,
                span: source_code.span_of(source_tag),
            };

            let data = parts.next().map(|p| {
                // this is a bit ugly
                // if xref was not present, there's two more splits...
                // re-slice the remainder of the string
                let span = line.span_of(p);
                let full_data = line.slice_from(span.offset());
                to_sourced(full_data)
            });

            Ok((
                level,
                Sourced {
                    span: source_code.span_of(line),
                    value: RawLine { tag, xref, data },
                },
            ))
        }();

        Some(result)
    })
}
