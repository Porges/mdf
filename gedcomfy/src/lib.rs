//! This is a library for parsing and validating GEDCOM files.

use core::str;
use std::borrow::Cow;

use encodings::{DataError, GEDCOMEncoding, MissingRequiredSubrecord};
use miette::SourceSpan;
use parser::{
    decoding::DecodingError,
    lines::{iterate_lines, LineSyntaxError},
    options::{OptionSetting, ParseOptions},
    records::RawRecord,
    GEDCOMSource, Sourced,
};
use vec1::Vec1;
use versions::GEDCOMVersion;

pub mod encodings;
pub mod highlighting;
mod ntypes;
pub mod parser;
pub(crate) mod v5;
pub(crate) mod v7;
pub(crate) mod versions;

#[derive(thiserror::Error, Debug, miette::Diagnostic)]
pub enum ValidationError {
    #[error("{} syntax error{} detected", .errors.len(), if .errors.len() > 1 { "s" } else { "" })]
    SyntaxErrorsDetected {
        #[related]
        errors: Vec1<LineSyntaxError>,
    },
    #[error("Encoding error detected: further validation errors will not be found")]
    #[diagnostic(transparent)]
    EncodingErrorDetected {
        #[from]
        error: DecodingError,
    },
}

impl<'a, S: GEDCOMSource + ?Sized> RawRecord<'a, S> {
    pub(crate) fn subrecord_optional(&self, subrecord_tag: &str) -> Option<&Sourced<RawRecord<S>>> {
        self.records
            .iter()
            .find(|r| r.value.line.tag.value == subrecord_tag)
    }

    pub(crate) fn subrecord(
        &self,
        subrecord_tag: &'static str,
        subrecord_description: &'static str,
    ) -> Result<&Sourced<RawRecord<S>>, MissingRequiredSubrecord> {
        self.subrecord_optional(subrecord_tag)
            .ok_or(MissingRequiredSubrecord {
                tag: subrecord_tag,
                description: subrecord_description,
            })
    }

    pub(crate) fn subrecords_optional(
        &self,
        tag: &'static str,
    ) -> impl Iterator<Item = &Sourced<RawRecord<S>>> {
        self.records
            .iter()
            .filter(move |r| r.value.line.tag.value == tag)
    }

    pub(crate) fn subrecords_required(
        &self,
        tag: &'static str,
        description: &'static str,
    ) -> Result<Vec1<&Sourced<RawRecord<S>>>, MissingRequiredSubrecord> {
        let v = Vec::from_iter(self.subrecords_optional(tag));
        Vec1::try_from(v).map_err(|_| MissingRequiredSubrecord { tag, description })
    }
}

/// Checks that the lines in the file are (minimally) well-formed.
/// Returns the number of lines in the file if successful.
pub fn validate_syntax(source: &[u8], buffer: &mut String) -> Result<usize, ValidationError> {
    validate_syntax_opt(source, buffer, ParseOptions::default())
}

pub fn validate_syntax_opt(
    source: &[u8],
    buffer: &mut String,
    parse_options: ParseOptions,
) -> Result<usize, ValidationError> {
    let (_version, source) = parser::decoding::detect_and_decode(source, parse_options)?;
    let source: &str = match source {
        Cow::Borrowed(input) => input,
        Cow::Owned(owned) => {
            *buffer = owned;
            buffer.as_str()
        }
    };

    let mut line_count = 0;
    let errors = Vec::from_iter(iterate_lines(source).filter_map(|r| match r {
        Ok(_) => {
            line_count += 1;
            None
        }
        Err(e) => Some(e),
    }));

    if let Ok(errors) = Vec1::try_from(errors) {
        Err(ValidationError::SyntaxErrorsDetected { errors })
    } else {
        Ok(line_count)
    }
}

pub(crate) struct FileFormatOptions {
    pub(crate) version_option: OptionSetting<GEDCOMVersion>,
    pub(crate) encoding_option: OptionSetting<GEDCOMEncoding>,
}

#[derive(thiserror::Error, Debug, miette::Diagnostic)]
pub(crate) enum FileStructureError {
    #[error("Missing HEAD record")]
    #[diagnostic(code(gedcom::schema_error::missing_head_record))]
    MissingHeadRecord {
        #[label("this is the first record in the file; the HEAD record should appear before it")]
        span: Option<SourceSpan>,
    },

    #[error("Missing trailer (TRLR) record")]
    #[diagnostic(
        code(gedcom::schema_error::missing_trailer_record),
        help("this record is always required at the end of the file â€“ GEDCOM file might be truncated?")
    )]
    MissingTrailerRecord,

    #[error("Records after trailer (TRLR) record")]
    #[diagnostic(
        code(gedcom::schema_error::records_after_trailer),
        help(
            "there are additional records after the trailer record which marks the end of the file"
        )
    )]
    RecordsAfterTrailer {
        #[label("this record appears after the TRLR record")]
        span: SourceSpan,
    },

    #[error(transparent)]
    #[diagnostic(transparent)]
    DataError(#[from] DataError<'static>),
}
