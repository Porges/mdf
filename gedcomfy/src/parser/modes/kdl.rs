use kdl::{KdlDocument, KdlEntry, KdlNode};

use crate::parser::{
    lines::LineValue, records::RawRecord, AnySourceCode, NonFatalHandler, ParseMode, ResultBuilder,
    Sourced,
};

#[derive(Default)]
pub(in crate::parser) struct Mode {}

impl NonFatalHandler for Mode {
    fn non_fatal<E>(&mut self, _error: E) -> Result<(), E>
    where
        E: Into<crate::parser::ParseError> + miette::Diagnostic,
    {
        Ok(())
    }
}

impl ParseMode for Mode {
    type ResultBuilder<'i> = Builder;

    fn get_result_builder<'i>(
        self,
        _version: crate::versions::SupportedGEDCOMVersion,
        _source_code: AnySourceCode,
    ) -> Result<Self::ResultBuilder<'i>, crate::parser::ParseError> {
        Ok(Builder {
            mode: self,
            doc: KdlDocument::new(),
        })
    }
}

pub(in crate::parser) struct Builder {
    mode: Mode,
    doc: KdlDocument,
}

impl NonFatalHandler for Builder {
    fn non_fatal<E>(&mut self, error: E) -> Result<(), E>
    where
        E: Into<crate::parser::ParseError> + miette::Diagnostic,
    {
        self.mode.non_fatal(error)
    }
}

impl<'i> ResultBuilder<'i> for Builder {
    type Result = KdlDocument;

    fn handle_record(
        &mut self,
        record: Sourced<RawRecord<'i>>,
    ) -> Result<(), crate::parser::ParseError> {
        self.doc.nodes_mut().push(record_to_kdl(record.value));
        Ok(())
    }

    fn complete(self) -> Result<Self::Result, crate::parser::ParseError> {
        Ok(self.doc)
    }
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.value.to_string()));
    }

    if let Some(mapped) = match record.line.line_value.value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children.nodes_mut().push(record_to_kdl(subrecord.value));
    }

    node.set_children(children);
    node
}
