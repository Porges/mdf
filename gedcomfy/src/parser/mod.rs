use std::borrow::Cow;

use ascii::{AsciiChar, AsciiStr};
use decoding::DecodingError;
use lines::iterate_lines;
use miette::{SourceOffset, SourceSpan};
use options::ParseOptions;
use records::{RawRecord, RecordBuilder};

pub mod encodings;
pub mod lines;
pub mod options;
pub mod records;

pub(crate) mod decoding;
pub(crate) mod versions;

/// Represents the minimal amount of decoding needed to
/// parse information from GEDCOM files.
pub trait GEDCOMSource: ascii::AsAsciiStr + PartialEq<AsciiStr> {
    fn lines(&self) -> impl Iterator<Item = &Self>;
    fn split_once(&self, char: AsciiChar) -> Option<(&Self, &Self)>;
    fn split_once_opt(&self, char: AsciiChar) -> (&Self, Option<&Self>) {
        match self.split_once(char) {
            Some((a, b)) => (a, Some(b)),
            None => (self, None),
        }
    }
    fn span_of(&self, source: &Self) -> SourceSpan;
    fn starts_with(&self, char: AsciiChar) -> bool;
    fn ends_with(&self, char: AsciiChar) -> bool;
    fn is_empty(&self) -> bool;
    fn slice_from(&self, offset: usize) -> &Self;
}

impl GEDCOMSource for str {
    fn lines(&self) -> impl Iterator<Item = &Self> {
        // GEDCOM lines are terminated by "any combination of a carriage return and a line feed"
        (*self).split(|c| c == '\r' || c == '\n').map(|mut s| {
            while s.starts_with('\n') || s.starts_with('\r') {
                s = &s[1..];
            }

            s
        })
    }

    fn span_of(&self, source: &Self) -> SourceSpan {
        debug_assert!(source.as_ptr() >= self.as_ptr());
        SourceSpan::new(
            SourceOffset::from(unsafe { source.as_ptr().byte_offset_from(self.as_ptr()) } as usize),
            source.len(),
        )
    }

    fn starts_with(&self, char: AsciiChar) -> bool {
        (*self).starts_with(char.as_char())
    }

    fn ends_with(&self, char: AsciiChar) -> bool {
        (*self).ends_with(char.as_char())
    }

    fn is_empty(&self) -> bool {
        (*self).is_empty()
    }

    fn slice_from(&self, offset: usize) -> &Self {
        &(*self)[offset..]
    }

    fn split_once(&self, char: AsciiChar) -> Option<(&Self, &Self)> {
        (*self).split_once(char.as_char())
    }
}

impl GEDCOMSource for [u8] {
    fn lines(&self) -> impl Iterator<Item = &Self> {
        // GEDCOM lines are terminated by "any combination of a carriage return and a line feed"
        (*self).split(|&x| x == b'\r' || x == b'\n').map(|mut s| {
            while s.starts_with(&[b'\n']) || s.starts_with(&[b'\r']) {
                s = &s[1..];
            }

            s
        })
    }

    fn span_of(&self, source: &Self) -> SourceSpan {
        debug_assert!(source.as_ptr() >= self.as_ptr());
        SourceSpan::new(
            SourceOffset::from(unsafe { source.as_ptr().byte_offset_from(self.as_ptr()) } as usize),
            source.len(),
        )
    }

    fn starts_with(&self, char: AsciiChar) -> bool {
        (*self).starts_with(&[char.as_byte()])
    }

    fn ends_with(&self, char: AsciiChar) -> bool {
        (*self).ends_with(&[char.as_byte()])
    }

    fn is_empty(&self) -> bool {
        (*self).is_empty()
    }

    fn slice_from(&self, offset: usize) -> &Self {
        &(*self)[offset..]
    }

    fn split_once(&self, char: AsciiChar) -> Option<(&Self, &Self)> {
        let ix = self.iter().position(|&x| x == char.as_byte())?;
        let (before, after) = self.split_at(ix);
        Some((before, &after[1..]))
    }
}

/// A value that is sourced from a specific location in a GEDCOM file.
///
/// This is used in many places to ensure that we can track back values
/// to their original location, which means that we can provide good
/// diagnostics in the case of errors.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct Sourced<T> {
    pub value: T,
    pub span: SourceSpan,
}

impl<T> Sourced<T> {
    pub(crate) fn as_ref(&self) -> Sourced<&T> {
        Sourced {
            value: &self.value,
            span: self.span,
        }
    }

    pub(crate) fn map<U>(self, f: impl FnOnce(T) -> U) -> Sourced<U> {
        Sourced {
            value: f(self.value),
            span: self.span,
        }
    }

    pub(crate) fn try_map<U, E>(self, f: impl FnOnce(T) -> Result<U, E>) -> Result<Sourced<U>, E> {
        Ok(Sourced {
            value: f(self.value)?,
            span: self.span,
        })
    }

    pub(crate) fn try_into<U>(self) -> Result<Sourced<U>, T::Error>
    where
        T: TryInto<U>,
    {
        match self.value.try_into() {
            Ok(value) => Ok(Sourced {
                value,
                span: self.span,
            }),
            Err(err) => Err(err),
        }
    }
}

impl<T, E> Sourced<Result<T, E>> {
    pub(crate) fn transpose(self) -> Result<Sourced<T>, E> {
        match self.value {
            Ok(value) => Ok(Sourced {
                value,
                span: self.span,
            }),
            Err(err) => Err(err),
        }
    }
}

/// A [`Sourced``] value derefs to the inner value, making
/// it easier to work with when the source information is not needed.
impl<T> std::ops::Deref for Sourced<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.value
    }
}

pub fn parse<'a>(
    input: &'a [u8],
    buffer: &'a mut String,
) -> Result<Vec<Sourced<RawRecord<'a>>>, DecodingError> {
    parse_opt(input, buffer, ParseOptions::default())
}

pub fn parse_opt<'a>(
    input: &'a [u8],
    buffer: &'a mut String,
    parse_options: ParseOptions,
) -> Result<Vec<Sourced<RawRecord<'a>>>, DecodingError> {
    let (_version, decoded_input) = decoding::detect_and_decode(input, parse_options)?;

    // if we want to return records without copying data in most cases,
    // the caller must provide a buffer where we can copy data if needed
    // (this happens when parsing UTF-16, or ANSEL if any non-ASCII characters are present)
    let input: &'a str = match decoded_input {
        Cow::Borrowed(borrowed) => borrowed,
        Cow::Owned(owned) => {
            *buffer = owned;
            buffer.as_str()
        }
    };

    let mut record_builder = RecordBuilder::new();

    let mut result = Vec::new();
    for line in iterate_lines(input) {
        match line {
            Ok(line) => {
                if let Some(record) = record_builder.handle_line(line)? {
                    result.push(record);
                }
            }
            Err(err) => return Err(err.into()),
        }
    }

    if let Some(record) = record_builder.complete()? {
        result.push(record);
    }

    Ok(result)
}
