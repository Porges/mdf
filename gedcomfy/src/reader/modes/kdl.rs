use kdl::{KdlDocument, KdlEntry, KdlNode};

use crate::reader::{
    lines::LineValue, records::RawRecord, NonFatalHandler, ReadMode, ResultBuilder, Sourced,
};

#[derive(Default)]
pub(in crate::reader) struct Mode {}

impl NonFatalHandler for Mode {
    fn report<E>(&mut self, _error: E) -> Result<(), E>
    where
        E: Into<crate::reader::ReaderError> + miette::Diagnostic,
    {
        Ok(())
    }
}

impl<'i> ReadMode<'i> for Mode {
    type ResultBuilder = Builder;

    fn into_result_builder(
        self,
        _version: crate::versions::SupportedGEDCOMVersion,
    ) -> Result<Self::ResultBuilder, crate::reader::ReaderError> {
        Ok(Builder {
            mode: self,
            doc: KdlDocument::new(),
        })
    }
}

pub(in crate::reader) struct Builder {
    mode: Mode,
    doc: KdlDocument,
}

impl NonFatalHandler for Builder {
    fn report<E>(&mut self, error: E) -> Result<(), E>
    where
        E: Into<crate::reader::ReaderError> + miette::Diagnostic,
    {
        self.mode.report(error)
    }
}

impl<'i> ResultBuilder<'i> for Builder {
    type Result = KdlDocument;

    fn handle_record(
        &mut self,
        record: Sourced<RawRecord>,
    ) -> Result<(), crate::reader::ReaderError> {
        self.doc
            .nodes_mut()
            .push(record_to_kdl(record.sourced_value));
        Ok(())
    }

    fn complete(self) -> Result<Self::Result, crate::reader::ReaderError> {
        Ok(self.doc)
    }
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.sourced_value.to_string()));
    }

    if let Some(mapped) = match record.line.value.sourced_value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children
            .nodes_mut()
            .push(record_to_kdl(subrecord.sourced_value));
    }

    node.set_children(children);
    node
}
