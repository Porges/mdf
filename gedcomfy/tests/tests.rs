use std::{path::PathBuf, sync::Once};

use gedcomfy::parser::{
    encodings::detect_external_encoding, lines::LineValue, options::ParseOptions, parse,
    records::RawRecord,
};
use kdl::{KdlDocument, KdlEntry, KdlNode};
use miette::{NamedSource, Report};

static INIT: Once = Once::new();
fn ensure_hook() {
    INIT.call_once(|| {
        miette::set_hook(Box::new(|_diag| {
            Box::new(
                miette::MietteHandlerOpts::new()
                    .terminal_links(false)
                    .unicode(true)
                    .color(false)
                    .width(132)
                    .build(),
            )
        }))
        .unwrap();
    });
}

#[test]
fn can_parse_allged_lines() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/others/allged.ged");

    let data = std::fs::read(path).unwrap();
    let mut buffer = String::new();

    let line_count = gedcomfy::validate_syntax(&data, &mut buffer).unwrap();
    assert_eq!(line_count, 1159);
}

#[test]
fn can_parse_allged_fully() -> miette::Result<()> {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/others/allged.ged");

    let parsed_file = gedcomfy::parse_file(&path, ParseOptions::default())?;
    insta::assert_debug_snapshot!(parsed_file);
    Ok(())
}

#[test]
fn produces_expected_allged_tree() {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/external/others/allged.ged");

    let data = std::fs::read(path).unwrap();
    let buffer = &mut String::new();
    let parsed = parse(&data, buffer).unwrap();

    insta::assert_snapshot!(to_kdl(parsed.into_iter().map(|r| r.value)));
}

#[test]
fn torture_test_valid() {
    ensure_hook();

    insta::glob!("external/torture-test-55-files/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        let buffer = &mut String::new();
        let parsed = parse(&data, buffer)
            .map_err(|e| Report::new(e).with_source_code(NamedSource::new(filename, data.clone())))
            .unwrap();
        insta::assert_snapshot!(to_kdl(parsed.into_iter().map(|r| r.value)));
    });
}

#[test]
fn golden_files() -> miette::Result<()> {
    ensure_hook();

    insta::glob!("syntax_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap();
        insta::with_settings!({
            // provide GEDCOM source alongside output
            description => String::from_utf8_lossy(&data),
        }, {
            match parse(&data, &mut String::new()) {
                Ok(records) => {
                    let kdl = to_kdl(records.into_iter().map(|r| r.value));
                    insta::assert_snapshot!(kdl);
                }
                Err(err) => insta::assert_snapshot!(format!(
                    "{:?}",
                    Report::new(err).with_source_code(NamedSource::new(filename.to_string_lossy(), data))
                )),
            };
        });
    });

    insta::glob!("format_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let filename = path.file_name().unwrap().to_string_lossy();
        insta::with_settings!({
            // provide GEDCOM source alongside output
            description => String::from_utf8_lossy(&data),
        }, {
            match parse(&data, &mut String::new()) {
                Ok(records) => {
                    let kdl = to_kdl(records.into_iter().map(|r| r.value));
                    insta::assert_snapshot!(kdl);
                }
                Err(err) => insta::assert_snapshot!(format!(
                    "{:?}",
                    Report::new(err).with_source_code(NamedSource::new(filename, data))
                )),
            };
        });
    });

    Ok(())
}

fn to_kdl<'a>(records: impl Iterator<Item = RawRecord<'a>>) -> KdlDocument {
    let mut doc = KdlDocument::new();
    for record in records {
        doc.nodes_mut().push(record_to_kdl(record));
    }

    doc
}

fn record_to_kdl(record: RawRecord) -> KdlNode {
    let mut node = KdlNode::new(record.line.tag.to_string());

    if let Some(xref) = &record.line.xref {
        node.entries_mut()
            .push(KdlEntry::new_prop("xref", xref.value.to_string()));
    }

    if let Some(mapped) = match record.line.line_value.value {
        LineValue::Ptr(None) => Some(KdlEntry::new_prop("see", kdl::KdlValue::Null)),
        LineValue::Ptr(Some(value)) => Some(KdlEntry::new_prop("see", value)),
        LineValue::Str(data) => Some(KdlEntry::new(data.to_string())),
        LineValue::None => None,
    } {
        node.entries_mut().push(mapped);
    }

    if record.records.is_empty() {
        return node;
    }

    let mut children = KdlDocument::new();
    for subrecord in record.records {
        children.nodes_mut().push(record_to_kdl(subrecord.value));
    }

    node.set_children(children);
    node
}

#[test]
fn test_encodings() {
    ensure_hook();

    insta::glob!("encoding_inputs/*.ged", |path| {
        let data = std::fs::read(path).unwrap();
        let data_ref: &[u8] = data.as_ref();
        let filename = path.file_name().unwrap().to_string_lossy();
        let encoding_report = match detect_external_encoding(data_ref) {
            Ok(Some(detected)) => format!(
                "External encoding detected: {}\nReason: {}",
                detected.encoding(),
                Report::new(detected.reason())
                    .with_source_code(NamedSource::new(filename.clone(), data.clone()))
            ),
            Ok(None) => "No external encoding detected (ASCII-compatible)".to_string(),
            Err(err) => format!(
                "{}",
                Report::new(err).with_source_code(NamedSource::new(filename.clone(), data.clone()))
            ),
        };

        insta::with_settings!({
            // provide GEDCOM source alongside output
            description => String::from_utf8_lossy(&data),
        }, {
            insta::assert_snapshot!(encoding_report);
            match parse(&data, &mut String::new()) {
                Ok(records) => {
                    let kdl = to_kdl(records.into_iter().map(|r| r.value));
                    insta::assert_snapshot!(kdl);
                }
                Err(err) => insta::assert_snapshot!(format!(
                    "{:?}",
                    Report::new(err).with_source_code(NamedSource::new(filename, data))
                )),
            };
        });
    });
}
