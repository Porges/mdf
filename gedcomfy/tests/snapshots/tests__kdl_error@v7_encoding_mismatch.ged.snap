---
source: gedcomfy/tests/tests.rs
description: "��0\u0000 \u0000H\u0000E\u0000A\u0000D\u0000\r\u0000\n\u00001\u0000 \u0000G\u0000E\u0000D\u0000C\u0000\r\u0000\n\u00002\u0000 \u0000V\u0000E\u0000R\u0000S\u0000 \u00007\u0000.\u00000\u0000\r\u0000\n\u00000\u0000 \u0000T\u0000R\u0000L\u0000R\u0000\r\u0000\n\u0000"
expression: "format!(\"{:?}\", miette::Report::new(err))"
input_file: gedcomfy/tests/encoding_inputs/v7_encoding_mismatch.ged
---
gedcom::encoding::version_encoding_mismatch

  × An error occurred while parsing the GEDCOM file
  ├─▶ Unable to determine encoding of GEDCOM file
  ╰─▶ GEDCOM version 7.0 requires the encoding to be UTF-8, but the file encoding was determined to be UTF-16 (little-endian)

Advice: gedcom::encoding_reason::bom

  ☞ this encoding was detected from the byte-order mark (BOM) at the start of the file
